zipline/assets/futures.py

```diff
@@ -0,0 +1,128 @@
+    """ Allows users to look up future contracts.
+
+    Parameters
+    ----------
+    trading_algorithm : TradingAlgorithm
+    root_symbol : str, optional
+    as_of_date : pandas.Timestamp
+    """
+
+    def __init__(self, trading_algorithm, root_symbol='', as_of_date=None):
+        self.root_symbol = root_symbol.upper()
+
+        # If an as_of_date is provided, self._as_of_date uses that
+        # value, otherwise None. This attribute backs the as_of_date property.
+        if as_of_date:
+            self._as_of_date = (TradingEnvironment.instance().normalize_date(
+                                as_of_date))
+        else:
+            self._as_of_date = None
+
+        # Maintain a reference to the TradingAlgorithm which holds this
+        # FutureChain instance, so we can get its datetime.
+        self._trading_algorithm = trading_algorithm
+
+        # Attribute to cache the most up-to-date chain, and the dt when it was
+        # last updated.
+        self._current_chain = []
+        self._last_updated = None
+
+        # Get the initial chain, since self._last_updated is None.
+        self._maybe_update_current_chain()
+
+    def __repr__(self):
+        if self._as_of_date:
+            return 'FutureChain(root_symbol=\'%s\', as_of_date=\'%s\')' % (
+                self.root_symbol, self.as_of_date)
+        else:
+            return 'FutureChain(root_symbol=\'%s\')' % self.root_symbol
+
+    def _get_datetime(self):
+        """
+        Returns the current datetime of the simulation.
+
+        Returns
+        -------
+        pandas.Timestamp
+            The datetime of FutureChain's TradingAlgorithm, if valid.
+            Otherwise uses the period start of the simulation.
+        """
+        try:
+            dt = self._trading_algorithm.get_datetime()
+        except AttributeError:
+            dt = self._trading_algorithm.sim_params.period_start
+
+        # If the algo does not yet have a datetime, use the period start
+        if dt is None:
+            dt = self._trading_algorithm.sim_params.period_start
+
+        return TradingEnvironment.instance().normalize_date(dt)
+
+    @property
+    def as_of_date(self):
+        """
+        The current as-of date of this future chain.
+
+        Returns
+        -------
+        pandas.Timestamp
+            The user-provided as_of_date if given, otherwise the
+            current datetime of the simulation.
+        """
+        if self._as_of_date is not None:
+            return self._as_of_date
+        else:
+            return self._get_datetime()
+
+    def _maybe_update_current_chain(self):
+        """ Updates the current chain if it's out of date.
+
+            Returns
+            -------
+            [Future]
+                The up-to-date current chain.
+        """
+        if (self._last_updated is None or
+           (self._last_updated != self._get_datetime())):
+
+            asset_finder = TradingEnvironment.instance().asset_finder
```

Comment:
Can we just pass in the asset_finder on construction rather than pulling it out of a global every time this is called?